### PR TITLE
core,grpclb: use denser atomics for census

### PR DESCRIPTION
**Pull request by carl-mastrangelo**
These are created per call, so use volatiles to save memory

### CODE REVIEW

File: grpclb/src/main/java/io/grpc/grpclb/GrpclbClientLoadRecorder.java [hunk 1]
```diff
@@ -36,44 +35,55 @@
...
+  private volatile long callsFinished;
 
   // Specific finish types
-  // Access to it should be protected by lock.  Contention is not an issue for these counts, because
-  // normally only a small portion of all RPCs are dropped.
-  @GuardedBy("this")
-  private HashMap<String, AtomicLong> callsDroppedPerToken = new HashMap<String, AtomicLong>();
-  private final AtomicLong callsFailedToSend = new AtomicLong();
-  private final AtomicLong callsFinishedKnownReceived = new AtomicLong();
+  private final ConcurrentMap<String, Long> callsDroppedPerToken =
```
Comment (ejona86) [inline grpclb/src/main/java/io/grpc/grpclb/GrpclbClientLoadRecorder.java]: Why this change?

I'm not excited to see a change like this among all the noise, without it being called out.

Comment (carl-mastrangelo) [reply grpclb/src/main/java/io/grpc/grpclb/GrpclbClientLoadRecorder.java]: Avoids locks, and avoid the extra atomics.  After talking with @zhangkun83 the reason to not use it is because Longs are less dense than AtomicLongs.

File: grpclb/src/main/java/io/grpc/grpclb/GrpclbClientLoadRecorder.java [hunk 2]
```diff
@@ -83,52 +100,55 @@ ClientStats generateLoadReport() {
...
+        .setNumCallsStarted(callsStartedUpdater.getAndSet(this, 0))
+        .setNumCallsFinished(callsFinishedUpdater.getAndSet(this, 0))
+        .setNumCallsFinishedWithClientFailedToSend(callsFailedToSendUpdater.getAndSet(this, 0))
+        .setNumCallsFinishedKnownReceived(callsFinishedKnownReceivedUpdater.getAndSet(this, 0));
+
+    Map<String, LongHolder> localCallsDroppedPerToken = Collections.emptyMap();
     synchronized (this) {
-      savedCallsDroppedPerToken = callsDroppedPerToken;
-      callsDroppedPerToken = new HashMap<String, AtomicLong>();
+      if (!callsDroppedPerToken.isEmpty()) {
```
Comment (ejona86) [inline grpclb/src/main/java/io/grpc/grpclb/GrpclbClientLoadRecorder.java]: In a noisy CL like this, a change like this which changes the logic is _really_ hard to notice. It should be separate or called out for reviewers.

This change itself looks fine, but it needs to be seen by reviewers.

File: grpclb/src/main/java/io/grpc/grpclb/GrpclbClientLoadRecorder.java (hunk 1 above)
Comment (ejona86) [reply grpclb/src/main/java/io/grpc/grpclb/GrpclbClientLoadRecorder.java]: Ah, I missed the fact it swapped from AtomicLong to Long. I mainly saw the @GuardedBy removal and the ConcurrentMap API usages.

File: grpclb/src/main/java/io/grpc/grpclb/GrpclbClientLoadRecorder.java (hunk 2 above)
Comment (carl-mastrangelo) [reply grpclb/src/main/java/io/grpc/grpclb/GrpclbClientLoadRecorder.java]: Fair enough.  I was hoping that the behavior was identical, which is why I didn't bring it up.
